broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+    // It could be just Receiver, but weirdly it doesn't implement futures_core Stream directly.
+    announce_rx: ReceiverStream<SubAnnounce>,
+    subs_to_all: Vec<Sender<SafekeeperTimelineInfo>>,
+    subs_to_timelines: HashMap<TenantTimelineId, Vec<SubSender>>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Publisher {
+    // Send msg to relevant subscribers.
+    pub fn send_msg(&mut self, msg: &SafekeeperTimelineInfo) -> Result<(), Status> {
+        // send message to subscribers for everything
+        let mut cleanup_subs_to_all = false;
+        for sub in self.subs_to_all.iter() {
+            match sub.try_send(msg.clone()) {
+                Err(TrySendError::Full(_)) => {
+                    trace!("dropping message, channel is full");
+                }
+                Err(TrySendError::Closed(_)) => {
+                    cleanup_subs_to_all = true;
+                }
+                _ => (),
+            }
+        }
+        // some channels got closed (subscriber gone), remove them
+        if cleanup_subs_to_all {
+            self.subs_to_all.retain(|tx| !tx.is_closed());
+        }
+
+        // send message to per timeline subscribers
+        let ttid =
+            parse_proto_ttid(msg.tenant_timeline_id.as_ref().ok_or_else(|| {
+                Status::new(Code::InvalidArgument, "missing tenant_timeline_id")
+            })?)?;
+        if let Some(subs) = self.subs_to_timelines.get(&ttid) {
+            for tx in subs.iter().map(|sub_sender| &sub_sender.1) {
+                if let Err(TrySendError::Full(_)) = tx.try_send(msg.clone()) {
+                    trace!("dropping message, channel is full");
+                }
+                // closed channel is ignored here; we will be notified and remove it soon
+            }
+        }
+        Ok(())
+    }
+
+    // Add/remove subscriber according to sub_announce.
+    pub fn update_sub(&mut self, sub_announce: SubAnnounce) {
+        match sub_announce {
+            SubAnnounce::AddAll(tx) => self.subs_to_all.push(tx),
+            SubAnnounce::AddTimeline(ttid, sub_sender) => {
+                match self.subs_to_timelines.entry(ttid) {
+                    Entry::Occupied(mut o) => {
+                        let subsenders = o.get_mut();
+                        subsenders.push(sub_sender);
+                    }
+                    Entry::Vacant(v) => {
+                        v.insert(vec![sub_sender]);
+                    }
+                }
+            }
+            SubAnnounce::RemoveTimeline(ref ttid, sub_id) => {
+                remove_sub(&mut self.subs_to_timelines, ttid, sub_id);
+            }
+        }
+    }
+}
+
+impl Drop for Publisher {
+    fn drop(&mut self) {
+        self.registry.unregister_publisher(self);
+    }
+}
+
+struct NeonBrokerImpl {
+    registry: Registry,
+}
+
+#[tonic::async_trait]
+impl NeonBroker for NeonBrokerImpl {
```

Comment:
NIT: Maybe call it the service `BrokerService`, and then struct becomes just `Broker`, so ends up `impl BrokerService for Broker`

Reply:
It would make imports more verbose (already fat), so let's leave as is.

Reply:
> It would make imports more verbose (already fat),

But why? As I can see there wont be any difference. And no ugly `Impl` suffix as a plus.

Reply:
tonic generated code includes two modules, currently called `neon_broker_server` and `neon_broker_client`, these would be longer. But actually I'm fine with either variant.

Reply:
Longer than what? There is a change from NeonBroker to BrokerService and struct can be called just Broker after that. So overall it looks shorter to me.

`neon_broker_proto` module name  is written in our code. So can be just proto. It is in lib.rs so can be either `crate::proto` or `broker::proto`. This is part of lengthy imports discussion above

Reply:
> neon_broker_proto module name is written in our code. So can be just proto. 

+1

> There is a change from NeonBroker to BrokerService and struct can be called just Broker after that. So overall it looks shorter to me.

Ok, in this case it won't be longer, but I dislike `BrokerService` names as it generates imports like
```
use neon_broker::proto::broker_service_server::{BrokerService, BrokerServiceServer};
```

So perhaps just `Broker`?

Reply:
pondering some more, made it `BrokerService`.

Reply:
There have to be two entities. One is a trait that is generated by tonic and the second one is the implementation struct. So cannot be Broker in both places (or we need to alias something in `use` section). So yeah, `BrokerService` trait and `Broker` struct is IMO is a good way to describe this